grails-app/controllers/org/pih/warehouse/api/ProductSupplierApiController.groovy

```diff
@@ -0,0 +1,16 @@
+package org.pih.warehouse.api
+
+import grails.converters.JSON
+import org.pih.warehouse.data.ProductSupplierService
+import org.pih.warehouse.product.ProductSupplierListDto
+import org.pih.warehouse.product.ProductSupplierListParams
+
+class ProductSupplierApiController {
+
+    ProductSupplierService productSupplierService
+
+    def list(ProductSupplierListParams filterParams) {
```

Comment:
I don't mind this, but the ProductSupplierListParams implementation seems to defeat the purpose of automatic data binding. 
```
class ProductSupplierListParams extends PaginationParams {
    String product
    String supplier
    ...
```
In other words, we don't want to bind to String, we want to auto-bind to objects like Product, Organization, etc to allow GORM to help us with the object lookup. 

If for some reason that wasn't working for you, then that means we don't have a proper Converter (previously known as a PropertyEditor) for that particular class. However, my understanding is that Grails 3+ has a default data binding mechanism that will take a command class like this
```
class ProductSupplierListCommand extends PaginationCommand {
    Product product
    Organization supplier
    ...
```
and bind the objects appropriatey.

Also, this is just a command object so just call it *Command unless we agree on a naming convention for these subtypes.

Reply:
Sure it would work the way you described, I just thought there is no sense to bind the `Product` and `Organization` if I eventually use only the id for filtering, e.g.:
```groovy
 if (params.product) {
    eq("product.id", params.product)
}
```
so I didn't see a sense to bind a whole product object if I eventually need only the id going further.

I appreciate you like the idea overall, because it's somewhat my own convention thrown at the deep end, as:
1. I didn't want to pass web params to the service layer, like we do in such methods now, because we had discussed this is not correct.
2. I didn't want to pass X amount of params to the service, because the method would look ugly (passing 10 params to a method)
3. By doing it in a "command way", we exactly know by looking at the code, what params user might provide and we don't have eventually to "debug" the web params in order to see, what is passed.
4. According to 3rd - we just have a full control of passed params and we know what will be binded - if a user makes mistake (e.g. he/she changes params in the URL), we don't proceed this param at all, because it won't be eventually binded to the command.


Let me know if my explanation about using `String` vs `Product`/`Organization` makes sense to you, or if I should try to check the equality by the object so:
```groovy
 if (params.product) {
    eq("product", params.product)
}
```
but the question is if we really need the whole object here? :thinking: 

As for naming - I named it like this just because it's somewhat new convention to the api list endpoints and wanted to distinguish them somehow.